src/audio/buffer.c

```diff
@@ -302,11 +309,17 @@ void buffer_attach(struct comp_buffer *buffer, struct list_item *head, int dir)
 	if (further_buffers_exist)
 		dcache_writeback_region(needs_sync, sizeof(struct list_item));
 	/* The cache line can be prefetched here, invalidate it after prepending */
-	list_item_prepend(buffer_comp_list(buffer, dir), head);
+	list_item_prepend(list, head);
 	if (further_buffers_exist)
 		dcache_invalidate_region(needs_sync, sizeof(struct list_item));
+	/* no dirty cache lines exist for this buffer yet, no need to write back */
+	dcache_invalidate_region(uncache_to_cache(list), sizeof(*list));
```

Comment:
really need this invalidate? should be automatic? if need, maybe a lot of other places need as well.

Reply:
> really need this invalidate? should be automatic? if need, maybe a lot of other places need as well.

@btian1 it's our current understanding that if memory is accessed via its uncached address, it still can be prefetched. Which isn't a problem if we continue accessing that memory only via that uncached address as is usually the case for us. But if you at some point do access that memory via cache - then you get a problem. This is the case here, therefore I add cache invalidation.

Reply:
ok, I am not 100% understand SOF's invalidate and writeback, normally, those should be done automatically by DSP itself.
Regarding the case you mentioned previously, when need send data from memory to DMA, then it need a writeback definitely,  or in the case memory is shared between different cores.

Just a little worry, with more and more invalidate and writeback, performance will get impacted.